src/Razor/src/Microsoft.AspNetCore.Razor.VSCode/src/RazorCodeLensProvider.ts

```diff
@@ -0,0 +1,130 @@
+ * ------------------------------------------------------------------------------------------ */
+
+import * as vscode from 'vscode';
+import { isRazorCSharpFile, getRazorDocumentUri } from './RazorConventions';
+import { RazorLanguageFeatureBase } from './RazorLanguageFeatureBase';
+import { LanguageKind } from './RPC/LanguageKind';
+import { CancellationToken } from 'vscode-jsonrpc';
+
+class RazorCodeLens extends vscode.CodeLens {
+    constructor(
+        range: vscode.Range,
+        public uri: vscode.Uri,
+        public document: vscode.TextDocument,
+        command?: vscode.Command) {
+
+        super(range, command);
+    }
+}
+
+export class RazorCodeLensProvider
+    extends RazorLanguageFeatureBase
+    implements vscode.CodeLensProvider {
+
+    public async provideCodeLenses(document: vscode.TextDocument, token: vscode.CancellationToken) {
+        try {
+            const razorDocument = await this.documentManager.getDocument(document.uri);
+            if (!razorDocument) {
+                return;
+            }
+
+            const csharpDocument = razorDocument.csharpDocument;
+
+            // Get all the code lenses that applies to our projected C# document.
+            const codeLenses = await vscode.commands.executeCommand<vscode.CodeLens[]>(
+                'vscode.executeCodeLensProvider',
+                csharpDocument.uri) as vscode.CodeLens[];
+
+            // Re-map the CodeLens locations to the original Razor document.
+            const remappedCodeLenses = new Array<vscode.CodeLens>();
+            for (const codeLens of codeLenses) {
+                const result = await this.serviceClient.mapToDocumentRange(
+                    LanguageKind.CSharp,
+                    codeLens.range,
+                    razorDocument.uri);
+                if (result) {
+                    const newCodeLens = new RazorCodeLens(result.range, razorDocument.uri, document, codeLens.command);
+                    remappedCodeLenses.push(newCodeLens);
+                } else {
+                    // This means this CodeLens was for non-user code. We can safely ignore those.
+                }
+            }
+
+            return remappedCodeLenses;
+
+        } catch (error) {
+            return [];
+        }
+    }
+
+    public async resolveCodeLens(codeLens: vscode.CodeLens, token: vscode.CancellationToken) {
+        if (codeLens instanceof RazorCodeLens) {
+            return await this.resolveRazorCodeLens(codeLens, token);
+        }
+    }
+
+    private async resolveRazorCodeLens(codeLens: RazorCodeLens, token: CancellationToken): Promise<vscode.CodeLens> {
+        // Initialize with default values.
+        codeLens.command = {
+            title: '',
+            command: '',
+            arguments: []
+        };
+
+        try {
+            const razorDocument = await this.documentManager.getDocument(codeLens.uri);
+            if (!razorDocument) {
+                return codeLens;
```

Comment:
Given that we initialized the codeLens command with default values what type of behavior would happen here?

Reply:
Without initializing it to empty, it will display `!!missing COMMAND!!` in place of `1 reference`.
With the initialization, it makes a little empty space where the code lens would exist but that space would be empty.